File: multiprogramming/Process.cpp
#include "Process.hpp"
using namespace std;

Process::Process() {

}

Process::Process (unsigned id, unsigned maxT, unsigned op, unsigned batch, int a, int b)
    : id(id), maxT(maxT), op(op), batch(batch), a(a), b(b), remainingT(maxT), elapsedT(0) {

}

unsigned Process::getPid() {
    return this->id;
}

unsigned Process::getBatch() {
    return this->batch;
}

unsigned Process::getMaxT() {
    return this->maxT;
}

unsigned Process::getRemainingT() {
    return this->remainingT;
}

unsigned Process::getElapsedT() {
    return this->elapsedT;
}

int Process::getOpt() {
    return this->op;
}

bool Process::getProcessed() {
    return this->processed;
}

int Process::getA() {
    return this->a;
}

int Process::getB() {
    return this->b;
}

string Process::getResult() {
    return to_string(calculateResult(a, b));
}

string Process::getOperator() {
    stringstream operation;

    operation << a << " " << defineOperator() << " " << b;

    return operation.str();
}

void Process::setMaxT(unsigned m){
    this->maxT=m;
}

void Process::setRemainingTime(unsigned r){
    this->remainingT=r;
}

void Process::setResult(string re){
    this->result=re;
}

int Process::calculateResult(int a, int b) {
    if (op == 1) {
        return a+b;
    } else if (op == 2) {
        return a-b;
    } else if (op == 3) {
        return a*b;
    } else if (op == 4) {
        return a/b;
    } else if (op == 5) {
        return a%b;
    } else {
        return 0;
    }
}

char Process::defineOperator() {
    if (op == 1) {
        return '+';
    } else if (op == 2) {
        return '-';
    } else if (op == 3) {
        return '*';
    } else if (op == 4) {
        return '/';
    } else if (op == 5) {
        return '%';
    } else {
        return ' ';
    }
}

void Process::setProcessed(bool p){
    this->processed=p;
}

string Process::toString() {
    stringstream details;

    details << "PID: " << id << endl;
    details << "Operation: " << a << " " << defineOperator() << " " << b << endl;
    details << "Max Time: " << maxT << endl;
    details << "Elapsed Time: " << elapsedT << endl;
    details << "Remaining Time: " << remainingT << endl;

    return details.str();
}

string Process::information() {
    stringstream info;
    info << id << "\t | " << maxT;
    return info.str();
}

string Process::completedInfo(){
    stringstream details;

    details << "PID: " << id << endl;
    details << "Operation: " << a << " " << defineOperator() << " " << b << endl;
    details << "Result: " << calculateResult(a, b) << endl;
    details << "Batch N.: " << batch << endl;

    return details.str();
}

int Process::updateTime() {
    this->remainingT--;
    this->elapsedT++;
    return remainingT;
}

vector<string> Process::vectorizeTask() {
    vector<string> result;

    result.push_back(to_string(id));                                // Process id
    result.push_back("Operation: " + to_string(a) + " " +           // Operation
                    defineOperator() + " " + to_string(b));
    result.push_back("Max Time: " + to_string(maxT));               // Max Expected Time
    result.push_back("Elapsed Time: " + to_string(elapsedT));       // Elapsed Time
    result.push_back("Remaining Time: " + to_string(remainingT));   // Remaining Time

    return result;
}

File: multiprogramming_py/Process.py
class Process:
    def __init__(self, pid=0, maxT=0, elapsedT=0, batch=0, a=0, b=0):
        self.pid = pid
        self.maxT = maxT
        self.elapsedT = elapsedT
        self.remainingT = maxT - elapsedT
        self.batch = batch
        self.a = a
        self.b = b
        self.op = ""
        self.result = ""

    # Class getters
    def get_pid(self):
        return self.id

    def get_maxT(self):
        return self.maxT

    def get_elapsedT(self):
        return self.elapsedT

    def get_remainingT(self):
        return self.remainingT

    def get_batch(self):
        return self.batch

    def get_result(self):
        return self.result

    def get_op(self):
        return self.op
    
    # Class Setters
    def set_maxT(self, maxT):
        self.maxT = maxT
        self.remainingT = maxT - self.elapsedT

    def set_elapsedT(self, elapsedT):
        self.elapsedT = elapsedT
        self.remainingT = self.maxT - elapsedT

    def set_batch(self, batch):
        self.batch = batch

    def set_a(self, a):
        self.a = a
        self.result = self.calculate_result(self.a, self.b)

    def set_b(self, b):
        self.b = b
        self.result = self.calculate_result(self.a, self.b)

    # Class methods
    def update_time(self):
        self.elapsedT += 1
        self.remainingT = self.maxT - self.elapsedT